sql/core/src/test/scala/org/apache/spark/sql/SQLContextSuite.scala

```diff
@@ -142,4 +144,86 @@ class SQLContextSuite extends SparkFunSuite with SharedSparkContext {
     }
   }
 
+  test("Bug SNAP-2061 Nested POJO object not handled when creating DataFrame from RDD") {
+    val sqlContext = SQLContext.getOrCreate(sc)
+    val personsCollection = for (k <- 1 until 100) yield {
+      new Person(k, "name_" + k, k.toLong, k.toShort,
+        k.toByte, k.toDouble *86.7543d, k.toFloat *7.31f,
+        true, Array.fill[Byte](k)(k.toByte),
+        new java.sql.Date(7836*k*1000), new Timestamp(7896*k*1000),
+        new Address("12320 sw horizon," + k, 97007*k))
+    }
+
+    // create a pair RDD from the collection
+    val personsRDD = sc.parallelize(personsCollection)
+    val df = sqlContext.createDataFrame(personsRDD, classOf[Person])
```

Comment:
Does the Encoder route (SparkSession.createDataset) work correctly for nested POJOs? Please add a test for that too. That is the more supported route since Spark 2.x